### PR TITLE
Introduce topo library 

### PR DESCRIPTION
**Pull request by anstaf**
- topological location classes `vertex`, `edge` and `cell`
 - notion of neighbor `chain` and the `link`of the `chain`
 - syntax sugar for producing chains. example: `vertex::cell::edge`.
 - predicates for locations, links and chains
 - an API on topological entities that is used in mesh concept and hides implementation details
 - test

### CODE REVIEW

File: include/gridtools/topo.hpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    using _impl::is_chain;
+    using _impl::is_link;
+    using _impl::is_location;
+
+    template <class Chain>
+    using reverse = typename _impl::reverse<Chain>::type;
+
+    template <class Chain>
+    using to = typename _impl::to<Chain>::type;
```
Comment (havogt) [inline include/gridtools/topo.hpp:170]: `to` and `from` are not intuitive for me. From the names I would expect that they construct a chain. Maybe `begin`/ `end` or `first`/`last`?

**Review by havogt — CHANGES_REQUESTED** [review]

**Comment by havogt** [comment]
I would leave it here.

File: include/gridtools/topo.hpp (hunk 1 above)
Comment (anstaf) [reply include/gridtools/topo.hpp]: done

File: tests/unit_tests/CMakeLists.txt [hunk 2]
```diff
@@ -34,6 +34,10 @@ endfunction()
 
 gridtools_check_compilation(test_meta test_meta.cpp)
 
+if (CMAKE_CXX_STANDARD GREATER_EQUAL 17)
```
[the remark this answers is not in the document]
Comment (havogt) [reply tests/unit_tests/CMakeLists.txt:37]: This is more a comment to myself. Let's think about the right solution.

File: include/gridtools/topo.hpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    template <class Chain>
+    using from = to<reverse<Chain>>;
+
+    // Takes the list of chains and join them together into a one.
+    // join<meta::list<vertex::edge, edge::vertex>> => vertex::edge::vertex
+    template <class Chains>
+    using join = meta::combine<meta::force<_impl::join>::apply, Chains>;
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/topo.hpp:179]: I extended the comment. Now it is documented that `join<meta::list<vertex::edge, vertex::edge>>` violates precondition.
Done.

File: tests/unit_tests/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (anstaf) [reply tests/unit_tests/CMakeLists.txt:37]: Sorry, it is still not clear to me what to do.
Can you expand your comment?

[the remark this answers is not in the document]
Comment (havogt) [reply tests/unit_tests/CMakeLists.txt:37]: After checking #1566, I think it's good.

**Review by havogt — APPROVED** [review]

**Comment by anstaf** [comment]
launch jenkins

**Comment by havogt** [comment]
tests ok

**Comment by anstaf** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

**Pull request merged**